File: 2021/go/day01.go
package main

import (
    "os"
    "bufio"
    "fmt"
    "strconv"
)

func get_input() []int {
    numbers := make([]int, 0)

    scanner := bufio.NewScanner(os.Stdin)

    for {
        scanner.Scan()
        s_n := scanner.Text()

        if len(s_n) == 0 { break }

        n, _ := strconv.Atoi(s_n)
        numbers = append(numbers, n)
    }

    return numbers
}

func solve_a(numbers []int) int {
    count := 0

    for i, n:= range numbers[:len(numbers) - 1] {
        if n < numbers[i + 1] {
            count += 1
        }
    }

    return count
}

func solve_b(numbers []int) int {
    avg_numbers := make([]int, len(numbers) - 2)
    for i, _:= range avg_numbers {
        avg_numbers[i] = numbers[i] + numbers[i + 1] + numbers[i + 2]
    }

    return solve_a(avg_numbers)
}


func main() {
    numbers := get_input()

    fmt.Println("Solution A-part", solve_a(numbers))
    fmt.Println("Solution B-part", solve_b(numbers))
}

File: 2021/rs/src/bin/day01.rs
mod day01 {
    pub fn solve_a(depths: &[i32]) -> usize {
        depths.iter()
            .zip(depths.iter().skip(1))
            .filter(|(d1, d2)| d1 < d2)
            .count()
    }

    pub fn solve_b(depths: &[i32]) -> usize {
        let smooth_depths: Vec<i32> = depths.iter()
            .zip(depths.iter().skip(1).zip(depths.iter().skip(2)))
            .map(|(d1, (d2, d3))| d1 + d2 + d3)
            .collect();

        solve_a(&smooth_depths)
    }
}

fn main() {
    use std::io::{self, BufRead};

    let stdin = io::stdin();

    let depths: Vec<i32> = stdin.lock().lines()
        .map(|d| d.unwrap().parse().unwrap())
        .collect();

    println!("Solution A-part: {}", day01::solve_a(&depths));
    println!("Solution B-part: {}", day01::solve_b(&depths));
}
